### PR TITLE
Implement remaining base functionality of ktxtools.

### PR DESCRIPTION
**Pull request by VaderY**
Briefs:
- Create: Implement --compare-psnr and --compare-ssim
- Create: Implement mipmap generation
- Create: Add *INT_PACK32 support
- Create: Add --raw create support for formats that are not part of the "format convertable" set
- Extract: Implement ASTC decoding for exporting
- Extract: Fix incorrect extension naming
- Encode: Implement --compare-psnr and --compare-ssim
- Help: Implement the help subcommand
- Tools: Change KTX Tools to link dynamically or statically based on the configuration
- Tools: Harden against invalid KTX files that could cause allocation failure
- Tools: General improvements and cleanup
- KTX/DFDUtils: Harden against invalid KTX files that could cause allocation failure
- KTX/DFDUtils: Add limits for printing garbage information from invalid KTX files
- CTS: Add additional tests
- CTS: Cleanup and organization
- Documentation: Extend and improve the markdown files in the repository
- Documentation: Add README.md to CTS that describes how CTS works with cross-references to the main KTX repo
- Documentation: Complete doxygen (manpage/html) documentation
- Documentation: Complete usage (--help) documentation

What will come later:
- Extract: Fragment-URI
- Tools: stdin support as input
- Additional support for 3D block-compressed, packed HDR, YUV and Depth/Stencil formats.
(These are the ones that were mostly missing from the KTX library)

Please review these changes and please also review the CTS repository https://github.com/KhronosGroup/KTX-Software-CTS/.

### CODE REVIEW

**Comment by VaderY** [comment]
Looks like the CI connection timeout troubles are over.
I have enabled --output-on-failure for ctest to have meaningful log if a test fails.
Currently on macOS x86_64 the ASTC encoding tests are failing and it is most likely caused by non-deterministic output. I will start the next week with looking into this.

**Comment by MarkCallow** [comment]
> Looks like the CI connection timeout troubles are over.

If this refers to the Chocolatey doxygen install issue, I added a retry loop. Since then I haven't seen the issue. I still see other connection issues from time to time on Travis.

> Currently on macOS x86_64 the ASTC encoding tests are failing and it is most likely caused by non-deterministic outputs. I will start the next week with looking into this.

Wait until I land PR #700. There have been changes to the default floating point settings in a recent clang version that cause differences with earlier versions. In that PR desired settings are explicitly set. If you have ASTC test images in the CTS repo you will have to regenerate them. With this PR the number of failures I see on arm64 vs x86_64 has shrunk.

**Comment by MarkCallow** [comment]
> Wait until I land PR #700.

Whoops! The issues I am pursuing in PR #700 have nothing to do with the failure you're seeing here. The version of the ASTC encoder currently in main and, presumably, therefore in this PR gives the same results for Linux, macOS and Windows on x86_64. The tests are all green so I do not understand what is happening here.

The issues being fixed in PR #700 are build issues with GCC 11 on Linux and different in recent versions of clang compared to earlier due to changes in default floating point options.

**Comment by MarkCallow** [comment]
There are a lot of warnings in the macOS build which will turn to errors when you rebase on main.

**Comment by VaderY** [comment]
>> Looks like the CI connection timeout troubles are over.

Last week there were multiple github outage that affect git and git lfs operations.

> If you have ASTC test images in the CTS repo you will have to regenerate them because this latest version of the encoder has many other changes compared with what we were using.

Golden file regeneration will not be an issue.

> The tests are all green so I do not understand what is happening here.

Yeah, it is strange. I will try a couple of things.

> There are a lot of warnings in the macOS build which will turn to errors when you rebase on main.

I will deal with any remaining warnings during/after we can rebase to main as some warnings were fixed there. But the newly added code should be warning free already.

File: tools/ktx/command_extract.cpp
Comment (MarkCallow) [inline tools/ktx/command_extract.cpp:104]: s/ based on the format save/, based on the format, save/

RAW is not an acronym so don't write in all caps. Use "Raw".

File: tools/ktx/command_extract.cpp
Comment (MarkCallow) [inline tools/ktx/command_extract.cpp:125]: What about depth/stencil? Explain their treatment. Are they not exported?

File: tools/ktx/command_extract.cpp
Comment (MarkCallow) [inline tools/ktx/command_extract.cpp:140]: Add a note saying block compressed transcode targets can only be saved in raw format.

**Review by MarkCallow — CHANGES_REQUESTED** [review]
Awesome work. I'm requesting many minor changes and 1 not so minor.

File: CMakeLists.txt
Comment (MarkCallow) [reply CMakeLists.txt:295]: We can leave this for now. After merge to main I'll work on using the Ninja Multi-Config generator in CI.

File: lib/info.c
Comment (VaderY) [reply lib/info.c:665]: If you mean the SGD it checks both byteOffset and byteLength. There is a line break due to length

File: lib/info.c
Comment (VaderY) [reply lib/info.c:847]: Hmm, everywhere else I was using constants, not sure how I left this one out, Fixed.

File: tools/ktx/command_help.cpp
Comment (VaderY) [reply tools/ktx/command_help.cpp:166]: It works either way. Both are explicit enough. I will remove it.

File: tools/ktx/command.h
Comment (VaderY) [reply tools/ktx/command.h:212]: Yes, cxxopts does include the default values in the usage string.

File: tools/ktx/command.h
Comment (VaderY) [reply tools/ktx/command.h:194]: This header file is in fact included by every command.

> if a command's output has some other focus it would not expose this option

Exactly, but "report output format" sounds good to me. Will change it.

File: tools/ktx/validate.cpp
Comment (VaderY) [reply tools/ktx/validate.cpp:450]: The current syntax is recognized by the IDEs (at least CLion, VSCode and vim handles it out of box).
But its trivial so I changed them all.

File: tools/ktx/command_create.cpp
Comment (VaderY) [reply tools/ktx/command_create.cpp:88]: The specification requested this. We can change it. Which one should be used?

Comment (VaderY) [reply tools/ktx/command_create.cpp:88]: `ktx encode --encode basis-lz ...` this one would look a bit odd
`ktx create --encode basis-lz ...`

`ktx encode --codec basis-lz ...`
`ktx create --codec  basis-lz ...`  this one would not be as explicit, but acceptable

File: tools/ktx/command_extract.cpp
Comment (VaderY) [reply tools/ktx/command_extract.cpp:125]: The complete support for depth/stencil formats were not part of this PR (as well as the other formats which needed extra work).

I have made an attempt to reword the rules:
- _UNORM formats (except Depth/Stencil) exported as PNG with RGB/RGBA 8/16 bit
- _SINT/_UINT formats exported as EXR with R/RG/RGB/RGBA Half/Float/UInt
- _SFLOAT/_UFLOAT formats exported as EXR with R/RG/RGB/RGBA Half/Float/UInt
- D16_UNORM exported as PNG with luminance (Gray) 16 bit
- Other Depth/Stencil formats exported as EXR with D/S/DS Half/Float

Not sure on what would be the best format to express these rules or weather we should just include the whole "format conversion table" or not.
I am not happy about this wording either. I am open to suggestions.

File: tools/ktx/command_create.cpp
Comment (VaderY) [reply tools/ktx/command_create.cpp:88]: The same is also true for 
`ktx extract --transcode` and 
`ktx transcode --target`

File: tools/ktx/compress_utils.h
Comment (VaderY) [reply tools/ktx/compress_utils.h:18]: I could not find a similar threshold for zlib. Included this for zstd.

File: tools/ktx/command_create.cpp
Comment (aqnuep) [reply tools/ktx/command_create.cpp:88]: I think the current names make better sense and reflect e.g. how git names the parameters when some other subcommand is used as part of the issued subcommand.

`ktx create --encode` is practically doing a similar job like the `ktx encode` subcommand as part of the create.

`ktx extract --transcode` is practically doing a similar job like the `ktx transcode` subcommand as part of the extract.

I don't necessarily agree that this needs to be changed, and I believe there was a reason to name them this way in the project specification.

File: tools/ktx/command_create.cpp
Comment (VaderY) [reply tools/ktx/command_create.cpp:116]: Possible, but the juice would not worth the squeeze. 
This text has to be present in the code anyways for the doxygen comments.

File: tools/ktx/command_create.cpp
Comment (VaderY) [reply tools/ktx/command_create.cpp:655]: I would vote for switching the word order in both flags:
```
--generate-mipmap
--runtime-mipmap
```

File: tools/ktx/metrics_utils.h
Comment (VaderY) [reply tools/ktx/metrics_utils.h:147]: The psnr functions only support unorm8. The ssim function is a float32 operation (with a unorm8 wrapper).
So their format support is very limited.

I dont think the math would allow ssim to be calculated per blocks. psnr might be doable (not a 100% on this), but not with this API.

> It is unfortunate we have to copy into the BasisU image class. We'll have to live with it for now.

Yeah, generic (template) library code would be the solution for this, but that is no possible. `std::mdspan` might help with this in the future to some degree, but even that would require support from the library. And even with mdspan a materialized unorm8 image data would be required which is not the case for most format.

File: tools/ktx/formats.h
Comment (VaderY) [reply tools/ktx/formats.h:98]: A multi index container would be overkill for this.

Option 1: Have a case insensitive string to VkFormat function generated
Option 2: Generate a single array of `pair<VkFormat, const char*>`s and write 2 wrapper function for the lookups (one direction could be binary searched and the other could be linear searched)
Option 3: Leave as is

That being said a proper system for VkFormats where you can query simple properties (without doing DFD operations) could replace most of `tools/ktx/formats.h`

**Comment by VaderY** [comment]
Thank you very much for the thorough review. It is greatly appreciated.

I made a pass on the review comments and I have pushed a commit for the ones that were marked as resolved.

Regarding the MacOS x86_64 ASTC error.
@MarkCallow are you aware of any issues or have any idea what could cause this mismatch?
Should we pursue this issue before we can rebase to the current main?
Should we disable these tests for this platform (similarly how we dealt with MSVC vs GCC/Clang determinism with BasisLZ) temporarily or permanently?

File: tools/ktx/command_create.cpp
Comment (VaderY) [reply tools/ktx/command_create.cpp:81]: I have changed the behavior and added extra tests to cover these.

**Comment by VaderY** [comment]
Most of the comments are now resolved. Only the ones that are waiting for further input are left.

File: lib/info.c
Comment (MarkCallow) [reply lib/info.c:666]: I know for SGD it checks both. My question is why doesn't it check byteOffset also for DFD and KVD?

File: tools/ktx/command_help.cpp
Comment (MarkCallow) [inline tools/ktx/command_help.cpp:30]: s/which/for which/

File: tools/ktx/encode_utils.h
Comment (MarkCallow) [reply tools/ktx/encode_utils.h]: You appear to have deleted the reference to ASTC encoding. Now at line 173 is reference to "ETC1S encoding" and "--encode etc1s". This needs to be fixed to match the actual terminology in use here.

File: tools/ktx/formats.h
Comment (MarkCallow) [reply tools/ktx/formats.h:98]: It is preferable if the function is generated so as Vulkan updates can be more easily handled though the likelihood of new formats must be decreasing over time.

I like 2 but as VkFormat to string is part of libktx and string to VkFormat is used by the tools, we're going to end up with 2 copies of the strings anyway.  It is also I expect more work to implement than 1.

Please do no. 1.

File: tools/ktx/command_create.cpp
Comment (MarkCallow) [reply tools/ktx/command_create.cpp]: Okay. Thanks for the analysis.

File: tools/ktx/command_create.cpp
Comment (MarkCallow) [reply tools/ktx/command_create.cpp:660]: Agree. Please make the change.

File: tools/ktx/command_create.cpp
Comment (MarkCallow) [reply tools/ktx/command_create.cpp:740]: Thanks. I see them now.

File: tools/ktx/metrics_utils.h
Comment (MarkCallow) [reply tools/ktx/metrics_utils.h:147]: Thank you. I am sure expanding on metrics and diff will come up in the future.

If ssim is a float32 operation why does that make "format support ... very limited"?

File: tools/ktx/command_create.cpp
Comment (MarkCallow) [reply tools/ktx/command_create.cpp:88]: Yes. `ktx encode --encode` would be odd. I'm persuaded. Leave as is.

File: tools/ktx/command_extract.cpp
Comment (MarkCallow) [reply tools/ktx/command_extract.cpp]: I think this wording is okay together with the preceding sentence (not shown here) about choosing the smallest related format.

**Comment by MarkCallow** [comment]
> Regarding the MacOS x86_64 ASTC error.
> @MarkCallow are you aware of any issues or have any idea what could cause this mismatch?
> Should we pursue this issue before we can rebase to the current main?
> Should we disable these tests for this platform (similarly how we dealt with MSVC vs GCC/Clang determinism with BasisLZ) temporarily or permanently?

Are the ASTC tests a new addition? If so, on what hardware and what compiler did you generated the golden files.

We don't have any issues with ASTC on x86_64 in main either pre or post PR #700. Tests pass on all platforms without any of them being disabled. The golden files were generated on macOS on x86_64 using code compiled with Xcode/Clang. I don't know the version pre PR #700. Probably Clang 13.x. The post PR #700 files used code compiled with Clang 14.

The issue we are having is that several of the tests fail when compiled for and run on an arm64 processor. This is true for Linux, macOS and Windows. This is clearly unrelated to the problem here.

We need to pursue this issue before rebasing so we understand it.

File: tools/ktx/metrics_utils.h
Comment (VaderY) [reply tools/ktx/metrics_utils.h:147]: What I meant by limited format support is that psnr is unorm8 only and even if ssnr supports float32 a format conversion is required for the input. I would also worry for formats like unorm32 where conversion to float32 is lossy.

File: tools/ktx/formats.h
Comment (MarkCallow) [reply tools/ktx/formats.h:98]: I will merge this without waiting for no. 1. Please do it in a subsequent PR.

File: tools/ktx/metrics_utils.h
Comment (MarkCallow) [reply tools/ktx/metrics_utils.h:147]: Thank you for the explanation.

File: tools/ktx/encode_utils.h
Comment (VaderY) [reply tools/ktx/encode_utils.h]: Fair point, we missed these reference changes, apologies. Hopefully they cleared up now.

File: lib/info.c
Comment (VaderY) [reply lib/info.c:666]: Oh I see. I checked to code and the spec, there is no reason for this discrepancy.
I have added the offset checks for DFD and KVD as there is no reason for attempting to print anything from byte 0.

File: tools/ktx/formats.h
Comment (VaderY) [reply tools/ktx/formats.h:98]: Will do. I would also want to allow format names with and without the VK_FORMAT_ prefix just like it is implemented for ktx tool for convenience. Is that okey?

**Comment by VaderY** [comment]
I misspoke, its ASTC decode and not encode.
Thank you for the info. I will try to check a couple of things and let you know about the results.

File: ci_scripts/on_failure.sh
Comment (MarkCallow) [inline ci_scripts/on_failure.sh:17]: For the long term we need something like the below tar to gather all the failing images and upload them. This means the failing images need to be easily identifiable by name or directory.

**Comment by MarkCallow** [comment]
> I misspoke, its ASTC decode and not encode.

Ahh! That explains why the existing toktx ASTC encode tests are all passing even though these tests are failing.

File: ci_scripts/on_failure.sh
Comment (VaderY) [reply ci_scripts/on_failure.sh:17]: This is just a quick and dirty just for this scenario. I think we could improve a CTS test runner to output the list of files that are mismatching. And based on that list they could be packed up.

The filepath of the output images matches their tests, so it is easy to find which test generated them.
For example all the `output/extract/png_2d/output_*.png` file come from the `tests/extract/png_2d.json` test.

**Comment by VaderY** [comment]
Success!

Looks like the problem was with the input files. I have created a new set of valid ASTC ktx2 files with valid ASTC data.
The previous data caused pixel to be magenta on MacOS and white on every other platform. I presume these color were representing invalid blocks or NaN related issues. I would have expected deterministic output even on invalid inputs, but I dont think that we should worry about this scenario.

File: ci_scripts/on_failure.sh
Comment (MarkCallow) [reply ci_scripts/on_failure.sh]: I like the idea of the runner outputting (writing to a file) the list of mismatched files. We need this and modified on_failure.{ps1,sh} soon. As CI builds take so long we can't afford to have to hack `on_failure` and re-run to capture mismatches. At least use of `--output-failure` removes the need to upload LastTest.log.

Please revert to the current `on_failure.sh` for now then I'll merge this. Plan on updating the CTS runner and scripts as part of the rebase.

**Comment by MarkCallow** [comment]
> The previous data caused pixel to be magenta on MacOS and white on every other platform.

White is never an error color. The spec is not clear on what error conditions are indicated by Magenta. 

"The error color is returned as an informative response to invalid conditions, including invalid block encodings or use of reserved endpoint modes."

I'd like to report this to ARM as it is their encoder and decoder we are using. Please provide one of the ktx2 files that gets the error and the input image that was encoded. Did you try to view one of the "bad" ktx2 files? to see if there is any magenta. That would be an interesting test because it would be using a different decoder.

I don't know what issue is causing the invalid certificate error. It keeps happening randomly. I've restarted the job.

**Comment by VaderY** [comment]
> Did you try to view one of the "bad" ktx2 files?

No I did not, I only used our decoders.

>Please provide one of the ktx2 files that gets the error and the input image that was encoded.

I will gather every related files and send you an email with the details.

> I don't know what issue is causing the invalid certificate error. It keeps happening randomly. I've restarted the job.

Thank you.
(Not sure is relevant, but once I run into an issue where github would reject clone requests when I tried to do more than 3 in parallel from the same IP. Not sure how the CI machines are setup.)

File: ci_scripts/on_failure.sh
Comment (VaderY) [reply ci_scripts/on_failure.sh]: I agree. The workarounds / debugging hacks have been already reverted yesterday.

**Review by MarkCallow — APPROVED** [review]

**Comment by MarkCallow** [comment]
> (Not sure is relevant, but once I run into an issue where github would reject clone requests when I tried to do more than 3 in parallel from the same IP. Not sure how the CI machines are setup.)

CI uses GitHub tokens to access GitHub. For each CI/repo pair the token is provided as a secure variable which means it is not available to PRs. Use of the token is for faster access. I think it has nothing to do with the intermittent certificate errors.

**Comment by MarkCallow** [comment]
I have just removed the webhook for notifying Appveyor of pushes, etc. The means no further Appveyor builds will be done. After rebasing to current `main` Windows builds will be done by GitHub Actions.

I removed the hook because all PR's based on `main` were showing Appveyor errors even though there is no .appveyor.yml.

**Pull request merged**
